// apollo-link-batch/dist/src/bundle.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.batchLink = {})));
}(this, (function (exports) { 'use strict';

var __extends = (undefined && undefined.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var apollo_link_core_1 = require("apollo-link-core");
var batching_1 = require("./batching");
var batching_2 = require("./batching");
exports.OperationBatcher = batching_2.OperationBatcher;
var BatchLink = (function (_super) {
    __extends(BatchLink, _super);
    function BatchLink(fetchParams) {
        var _this = _super.call(this) || this;
        _this.batchInterval = (fetchParams && fetchParams.batchInterval) || 10;
        _this.batchMax = (fetchParams && fetchParams.batchMax) || 0;
        if (typeof _this.batchInterval !== 'number') {
            throw new Error("batchInterval must be a number, got " + _this.batchInterval);
        }
        if (typeof _this.batchMax !== 'number') {
            throw new Error("batchMax must be a number, got " + _this.batchMax);
        }
        _this.batcher = new batching_1.OperationBatcher({
            batchInterval: _this.batchInterval,
            batchMax: _this.batchMax,
            batchHandler: fetchParams.batchHandler,
        });
        return _this;
    }
    BatchLink.prototype.request = function (operation, forward) {
        return this.batcher.enqueueRequest({
            operation: operation,
            forward: forward,
        });
    };
    return BatchLink;
}(apollo_link_core_1.ApolloLink));
exports.default = BatchLink;

Object.defineProperty(exports, '__esModule', { value: true });

})));
